### PR TITLE
Fix git dependencies

### PR DESCRIPTION
**Pull request by simonepri**
Resolves: https://github.com/python-poetry/poetry/issues/2326, https://github.com/huggingface/transformers/issues/3982

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: poetry/core/utils/helpers.py [hunk 1]
```diff
@@ -81,6 +81,8 @@ def parse_requires(requires):  # type: (str) -> List[str]
             continue
 
         if current_marker:
+            if "@" in line:
```
Comment (simonepri) [inline poetry/core/utils/helpers.py:84]: Is that sufficient to check whether it is a URL requirement?

File: poetry/core/utils/helpers.py [hunk 2]
```diff
@@ -81,6 +81,8 @@ def parse_requires(requires):  # type: (str) -> List[str]
...
+                line += " "
             line = "{}; {}".format(line, current_marker)
```
Comment (abn) [inline poetry/core/utils/helpers.py:86]:
```suggestion
        if current_marker:
            line = "{} ; {}".format(line, current_marker)
```

I wonder if this could introduce some edge case. I'd just add the whitespace in all cases.

File: tests/conftest.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 import pytest
 
 from poetry.core.utils._compat import Path
-from tests.utils import tempfile
+from tests._helpers.utils import tempfile
```
Comment (abn) [inline tests/conftest.py:6]: Probably just move utils.py to `tests/utils/__init__.py`

File: tests/packages/test_main.py [hunk 4]
```diff
@@ -183,6 +183,22 @@ def test_dependency_from_pep_508_with_git_url():
...
+        ' ; extra == "foo;"'
+    )
+
+    dep = dependency_from_pep_508(name)
+
+    assert "poetry" == dep.name
+    assert dep.is_vcs()
+    assert "git" == dep.vcs
+    assert "https://github.com/python-poetry/poetry.git" == dep.source
+    assert "b;ar;" == dep.reference
```
Comment (abn) [inline tests/packages/test_main.py:198]: Is this a valid scenario?

Comment (simonepri) [reply tests/packages/test_main.py:198]: Yes, you can habe a branch with `;` in the name.
See https://github.com/python-poetry/core/blob/fbfcda73b2c5524c3eda29b1d9b70734a3a9dcf3/tests/packages/test_vcs_dependency.py#L55

File: tests/conftest.py (hunk 3 above)
Comment (simonepri) [reply tests/conftest.py:6]: Hiding test helpers inside the `__init__.py` does not seem ideal to me.
Is that common?

An alternative that I like more, is to just rename it to something else like `testutils`

File: poetry/core/utils/helpers.py (hunk 2 above)
Comment (simonepri) [reply poetry/core/utils/helpers.py:86]: If we always add the space we break the old `parse_requires` tests (because the space was not expected).

Comment (abn) [reply poetry/core/utils/helpers.py:86]: I am okay with the test's expected output needing updates as long it won't break any existing use cases. As far as I can tell, this change shouldn't affect anything reading metadata we generate.

Comment (simonepri) [reply poetry/core/utils/helpers.py:86]: No it shouldn't, but I preferred to stick with the same representation generated here:
https://github.com/python-poetry/core/blob/e1c6978c5a0c104d373d78bcfc10a722409de583/poetry/core/_vendor/packaging/requirements.py#L123-L141

File: poetry/core/packages/__init__.py [hunk 5]
```diff
@@ -63,8 +63,8 @@ def dependency_from_pep_508(
     name = parts[0].strip()
     if len(parts) > 1:
         rest = parts[1]
-        if ";" in rest:
-            name += ";" + rest.split(";", 1)[1]
+        if " ;" in rest:
+            name += " ;" + rest.split(" ;", 1)[1]
```
Comment (abn) [inline poetry/core/packages/__init__.py:67]:
```suggestion
        if ";" in rest:
            name += " ;" + rest.split(";", 1)[1].strip()
```

File: poetry/core/utils/helpers.py (hunk 2 above)
Comment (abn) [reply poetry/core/utils/helpers.py:86]: Understandable; but this use case is a bit different as the objective here is to convert a requirement txt file into a list of requires_dist. And I suspect using `@` to check if url is not really reliable. Considering that the extra whitespace is good hygene in all cases, I'd prefer that we applied the above suggestion.

File: poetry/core/packages/__init__.py (hunk 5 above)
Comment (simonepri) [reply poetry/core/packages/__init__.py:67]: This won't work if the url contains `;`.

File: poetry/core/utils/helpers.py (hunk 2 above)
Comment (simonepri) [reply poetry/core/utils/helpers.py]: Done

Comment (simonepri) [reply poetry/core/utils/helpers.py]: Done

**Comment by simonepri** [comment]
@abn Any update on this?

**Comment by abn** [comment]
I will get to this one this week hopefully. I wanted to go through the relevant peps before getting back to you.

**Comment by tholiao** [comment]
This PR fixes an issue with a widely used library for natural language processing; there are two major conference deadlines in a month so any expedition would be greatly, greatly appreciated!

**Comment by simonepri** [comment]
@abn Sorry for pinging you again.
Do we have any blocker for this?
